### ISSUE TITLE
customized loss function as function of other columns of data?

### ISSUE THREAD

**Issue by phunterlau** [dmlc/xgboost#550]
In the recent finished Kaggle Flavour of Physics competition, the `UGradientBoostingClassifier` https://arogozhnikov.github.io/hep_ml/gb.html had some good performance using `BinFlatnessLossFunction` which is a function of the data column `'mass'`, the sample code is here https://www.kaggle.com/benhamner/flavours-of-physics/flatness-boosting-example/code However,`UGradientBoostingClassifier` is like the old GBM implementation running on single thread. Is it possible to implemented a loss function like `BinFlatnessLossFunction` with depends on some other columns in the same dataset, and use xgboost as the classifier for high speed? It would be very useful for high energy physics research. The `BinFlatnessLossFunction` is here https://arogozhnikov.github.io/hep_ml/losses.html#hep_ml.losses.BinFlatnessLossFunction

**Comment by tqchen** [comment]
As you know python function can be closures. So I think it is readily supported by current customized loss?

**Comment by tqchen** [comment]
The general idea
```python
other_data = col["mass"]

def grad(label, grad):
      # can use other_data here

pass grad to as customized loss
```

**Comment by phunterlau** [comment]
Good idea. Let me see if I can implement it, or some one else has interest too. This boosting loss function idea comes from this paper http://arxiv.org/pdf/1410.4140v1.pdf and practically useful for some cases. It would be a good demo for promoting xgboost.

**Comment by yenchenlin1994** [comment]
Hello @phunterlau,

Did you start to implement it?
I'm interested in working on this issue.

**Comment by phunterlau** [comment]
@yenchenlin1994 good to know that you can work on it. I haven't had a real chance of taking actions on it. Please let me know if you discover anything about it.
